x/evm/state/state.go

```diff
@@ -0,0 +1,127 @@
+package state
+
+import (
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+)
+
+func (s *StateDBImpl) CreateAccount(acc common.Address) {
+	// clear any existing state but keep balance untouched
+	s.clearAccountState(acc)
+
+	s.created[acc.String()] = struct{}{}
+	delete(s.selfDestructedAccs, acc.String())
+}
+
+func (s *StateDBImpl) GetCommittedState(addr common.Address, hash common.Hash) common.Hash {
+	return s.getState(addr, hash, func(kv sdk.KVStore, h common.Hash) []byte { return kv.GetCommitted(h[:]) })
+}
+
+func (s *StateDBImpl) GetState(addr common.Address, hash common.Hash) common.Hash {
+	return s.getState(addr, hash, func(kv sdk.KVStore, h common.Hash) []byte { return kv.Get(h[:]) })
+}
+
+func (s *StateDBImpl) getState(addr common.Address, hash common.Hash, getter func(sdk.KVStore, common.Hash) []byte) common.Hash {
+	val := getter(s.prefixStore(addr), hash)
+	if val == nil {
+		return common.Hash{}
+	}
+	return common.BytesToHash(val)
+}
+
+func (s *StateDBImpl) SetState(addr common.Address, key common.Hash, val common.Hash) {
+	s.prefixStore(addr).Set(key[:], val[:])
+}
+
+func (s *StateDBImpl) GetTransientState(addr common.Address, key common.Hash) common.Hash {
+	if addrState, ok := s.transientStorage[addr.String()]; !ok {
+		return common.Hash{}
+	} else if val, ok := addrState[key.String()]; !ok {
+		return common.Hash{}
+	} else {
+		return common.BytesToHash(val)
+	}
+}
+
+func (s *StateDBImpl) SetTransientState(addr common.Address, key, value common.Hash) {
+	addrKey := addr.String()
+	if addrState, ok := s.transientStorage[addrKey]; !ok {
+		s.transientStorage[addrKey] = map[string][]byte{key.String(): value[:]}
+	} else {
+		addrState[key.String()] = value[:]
+	}
+}
+
+// burns account's balance
+// clear account's state except the transient state (in Ethereum transient states are
+// still available even after self destruction in the same tx)
+func (s *StateDBImpl) SelfDestruct(acc common.Address) {
+	var balance sdk.Coin
+	if seiAddr, ok := s.k.GetSeiAddress(s.ctx, acc); ok {
+		// send all useis from seiAddr to the EVM module
+		balance = s.k.BankKeeper().GetBalance(s.ctx, seiAddr, s.k.GetBaseDenom(s.ctx))
+		if balance.Amount.Int64() != 0 {
```

Comment:
Is it possible for `balance` to be larger than `2^64`

Reply:
total Sei supply is 10^10, or 10^16 usei, whereas int64 can represent up to 9^18, so i think we should be good when it's just representing the absolute value of certain balance. We do need to check for overflow though when addition happens, just in case EVM hypothetically adds large sum onto some balances.